docs: Add design document, proto and tooling

juliaogris opened the pull request:
Add a design document and a README file to the repository documentation.

The design document provides an overview of Telejob, a remote job execution
service, following Teleport's system engineering challenge requirements and
RFD guidelines. It covers:

- Overview
- UX CLI flags
- API
- Security
- Process Execution Lifecycle
- Resource Management
- Log Streaming
- Design Approach
- Scope and Limitations

The README serves as the repository's front page, with a brief overview, a
link to the design document, and development instructions.

Add the `telejob.proto` file to the `proto` directory, defining the gRPC
service.

Integrate automated linting and formatting into the CI with `buf` using a GitHub
Actions Workflow. Define proto linting rules in `proto/buf.yaml`.

Use `just` for consistent and reproducible builds in both CI and local
development environments.

Manage tooling with Hermit to ensure consistent versions.

juliaogris committed:
tools: Manage tooling with hermit

Use hermit to manage repository tooling in isolated environments
(except for git). This improves reproducibility and consistency for all
contributors.

Initialize with:

	hermit init
	hermit buf just
	echo "/bin/* linguist-generated=true" > .gitattributes

Link: https://cashapp.github.io/hermit/

juliaogris committed:
proto: Add telejob.proto and tooling

Add the `telejob.proto` file to the `proto` directory, defining the gRPC
service.

Integrate automated linting and formatting into the CI pipeline with
`buf`. Define proto linting rules in `proto/buf.yaml`.

Use `just` for consistent and reproducible builds in both CI and local
development environments.

juliaogris committed:
docs: Add design document and README

Add a design document and a README file to the repository documentation.

The design document provides an overview of Telejob, a remote job
execution service, following Teleport's system engineering challenge
requirements and RFD guidelines. It covers:

- UX CLI flags
- API
- Security
- Process Execution Lifecycle
- Resource Management
- Log Streaming
- Design Approach
- Scope and Limitations

The README serves as the repository's front page, with a brief overview,
a link to the design document, and development instructions.

Link: https://github.com/gravitational/careers/blob/4a037eaac492f3ae94bf2bdbd6ef1685bd4842e9/challenges/systems/challenge-1.md
Link: https://github.com/gravitational/teleport/blob/aaf8cd94001552ad53679a87d5038ce5d2a5e03f/rfd/0000-rfds.md

smallinsky commented on docs/design-doc.md:
```diff
@@ -0,0 +1,301 @@
+8. **Asynchronous Wait:** A separate goroutine is launched to wait for the
+   command to terminate using `cmd.Wait()`. This allows the server to handle
+   other requests concurrently.
+9. **Job Termination (Optional):** If a user requests to stop the job, the
+   server sends a `SIGKILL` signal to the process. For simplicity, a more graceful
+   termination sequence (e.g., `SIGTERM`, `SIGKILL` with grace periods)
+   is not implemented.
```
Will sending a /SIGKILL to the parent process also terminate the child process in the case of a user job process being forked?

juliaogris replied:
Good point, forked child processes won't be killed with `cmd.Process.Kill()`.

It seems that writing `1` to the job's cgroup file `cgroup.kill` will kill the process and all child processes. So I will use that instead. I've updated the docs accordingly.

From the Kernel docs of `cgroup.kill`

> Writing “1” to the file causes the cgroup and all descendant cgroups to be killed. This means that all processes located in the affected cgroup tree will be killed via SIGKILL.

https://docs.kernel.org/admin-guide/cgroup-v2.html#core-interface-files

smallinsky commented on proto/telejob.proto:
```diff
@@ -0,0 +1,84 @@
+
+// StartRequest contains the command and arguments to execute.
+message StartRequest {
+  repeated string command = 1;
```
Does it mean that multiple commands will be executed at once?

> // StartRequest contains the command and arguments to execute.

Does it make sense to decouple the command from its arguments?

juliaogris replied:
OK, I have updated it to 
```proto
message StartRequest {
  string command = 1;
  repeated string arguments = 2;
}
```

I intended for `repeated string command` to mean command _and_ args
for simplicity, but I see from your question that this could be confusing.
Some people may think it means many commands.

smallinsky commented on proto/telejob.proto:
```diff
@@ -0,0 +1,84 @@
+// State represents the current state of a job, running or stopped.
+enum State {
+  STATE_UNSPECIFIED = 0;
+  STATE_RUNNING = 1;
+  STATE_STOPPED = 2;
```
How can a user determine whether a command was stopped, exited, or failed?

juliaogris replied:
I've added the following to the design doc:

> ### Termination Status
> 
> The `Status.ExitCode` field in the status response provides information about
> the job's termination status:
>
>- **Successful termination:** If the command runs to completion, the
>  `ExitCode` reflects the process's exit code (0-255).
>- **Termination by signal:** If the command is terminated by a signal, the
>  `ExitCode` is set to -1. That's the case when the user stops it. In the
>  client CLI, this is presented as `EXIT: signal` (see godocs of
>  [`ProcessState.ExitCode`] for details).
>- **Startup failure:** If the command cannot be started (e.g., due to a
>  non-existent command), the gRPC method immediately returns an error with
>  [`InvalidArgument` code]. In this case, the job is not created, and no
>  `ExitCode` is provided.
>- **Running job:** A running job has an `ExitCode` of -2, equivalent to the
>  `job.NotTerminated` constant.
>
>[`ProcessState.ExitCode`]: https://pkg.go.dev/os#ProcessState.ExitCode
>[`InvalidArgument` code]: https://pkg.go.dev/google.golang.org/grpc/codes#Code


However, it will not be possible to differentiate between a Job that has been stopped by the user and a job that was terminated elsewhere by a signal. In my opinion that's acceptable.

smallinsky commented on proto/telejob.proto:
```diff
@@ -0,0 +1,84 @@
+// ListRequest specifies whether to query all or only running jobs for
+// requesting user.
+message ListRequest {
+  bool all = 1;
```
The difference between all = false and all = true is somewhat clearer from the proto definition.

The list endpoint is not challenge-required, so please feel free to reduce the scope.

juliaogris replied:
I have removed the `list` endpoint altogether.

smallinsky commented on docs/design-doc.md:
```diff
@@ -0,0 +1,301 @@
+  rpc Stop(StopRequest) returns (StopResponse) {}
+  rpc List(ListRequest) returns (ListResponse) {}
+  rpc Status(StatusRequest) returns (StatusResponse) {}
+  rpc Logs(LogsRequest) returns (stream LogsResponse) {} // Added for manual testing convenience.
```
> // Added for manual testing convenience.

Could you elaborate? Streaming is one of the challenge requirements:

> Library should be able to stream the output of a running job.

This comment is confusing - How will this endpoint be used specifically for manual testing?

juliaogris replied:
The comment was intended for the `List` method, I mistakenly added it to `Logs`.
I have deleted the comment _and_ the `List` method.

smallinsky reviewed (commented):
I have left a few questions/suggestions but overall the design document looks neat.

nklaassen commented on docs/design-doc.md:
```diff
@@ -0,0 +1,301 @@
+
+Telejob uses a gRPC API secured with mutual TLS (mTLS) to facilitate
+communication between the server and client. The server manages job execution
+through the endpoints `start`, `stop`, `status`, `logs`, and `list`, while the
```
we're not asking for a `list` endpoint, while it would undoubtably be useful in a real system, we usually ask candidates to skip it just to cut scope a bit.

juliaogris replied:
Done, I have removed the `list` endpoint.

nklaassen commented on docs/design-doc.md:
```diff
@@ -0,0 +1,301 @@
+    list                       Print the status of the user's currently running jobs.
+        -a, --all              Include stopped jobs in the output.
+
+    logs <JOB_ID>              Print the logs of the job with the given <JOB_ID>.
+        -f, --follow           Continuously stream the log output.
```
you can just make -f the default/only option to cut scope a bit

juliaogris replied:
Done, I have removed the `--follow` flag and updated the `logs` description.

nklaassen commented on docs/design-doc.md:
```diff
@@ -0,0 +1,301 @@
+export TELEJOB_CLIENT_KEY=certs/client1.key
+export TELEJOB_SERVER_CA_CERT=certs/server-ca.crt
+
+# Test `start`, `stop`, `status`, and `list` commands.
+telejob start echo "Hello, world!"
+telejob start sleep 10
+telejob list --all
+telejob stop <JOB_ID>
+telejob status <JOB_ID>
```
could you include some example output here?

juliaogris replied:
Done, I have included some example output.

nklaassen commented on docs/design-doc.md:
```diff
@@ -0,0 +1,301 @@
+  - `TLS_AES_128_GCM_SHA256`
+  - `TLS_AES_256_GCM_SHA384`
+  - `TLS_CHACHA20_POLY1305_SHA256`
+- Cipher preference: client chooses
```
nit: assuming you are using Go's `crypto/tls`, it takes over cipher suite preference ordering, see the docs on https://pkg.go.dev/crypto/tls#Config, `PreferServerCipherSuites` is deprecated and ignored

juliaogris replied:
Thank you! I have  removed the "client chooses" line.

nklaassen commented on docs/design-doc.md:
```diff
@@ -0,0 +1,301 @@
+
+## Security
+
+### TLS setup
```
In this section I would like to see some description of your CA setup, what each cert is signed by and what it's trusted by

juliaogris replied:
Done, I have added an extra section. Please let me know if you still feel that it doesn't cover the CA setup, signing and trusting sufficiently.

nklaassen commented on docs/design-doc.md:
```diff
@@ -0,0 +1,301 @@
+
+### Authentication
+
+Telejob uses client certificates to verify user identities:
```
you described how the server authenticates the client, and how does the client authenticate the server?

juliaogris replied:
Server authentication is covered in the new section "Server Certificate".
Please let me know if this is not sufficient.

nklaassen replied:
looks great, thanks!

nklaassen commented on docs/design-doc.md:
```diff
@@ -0,0 +1,301 @@
+Telejob uses client certificates to verify user identities:
+
+1. The client presents its certificate when connecting.
+2. The server checks if the certificate is valid and trusted.
```
what does it mean to be valid and trusted?

juliaogris replied:
I have expanded this item to:

> 2. The server checks if the certificate is valid and trusted
>    - Verifies the digital signature.
>    - Ensures the certificate is within its valid time range.
>    - Confirms the presence of a `commonName`.
>    - Verifies the certificate is signed by the Client CA.

Please let me know if you think it needs further clarification.

nklaassen commented on docs/design-doc.md:
```diff
@@ -0,0 +1,301 @@
+- **Concurrent Client Support:** Multiple clients can simultaneously stream logs
+  from the same job without interference.
+
+The core of the log streaming functionality is the `logDispatcher`. It manages
+incoming log data from a job's standard output and standard error streams. This
+data is received through a dedicated input channel. To serve log streaming
+requests from multiple clients, the `logDispatcher` employs a broadcast
+(fan-out) approach, distributing log data over multiple output channels. This
+allows several clients to concurrently receive log streams from the same job.
```
Can you say something about how clients will "catch up" on output that was written before they attached to the logDispatcher fan-out mechanism? And will this design tolerate two clients where one reads very slowly and another reads very quickly?

juliaogris replied:
I have added a section "Log Streaming Implementation" where I go into more detail.
The following paragraph is most relevant to your questions but it may be necessary 
to read the full "Log Streaming Implementation" section.

> When a new reader requests log data, the `logDispatcher` first sends all
> historical logs. A log request contains a response channel and a start index
> holding the current position in the full log (initially 0). The reader **must**
> immediately read from the response channel and then process the log data, which
> may be time-consuming. Once processing is complete, the reader sends a
> follow-up request with an updated start index. If no new log data is available,
> the reader's response channel is registered in the `logDispatcher`'s followers
> list.

juliaogris replied:
This is roughly the algorithm I have in mind, excluding error handling and termination sequence:
```go
type logResponseCh chan []byte

type logRequest struct {
	startIdx int
	ch       logResponseCh
}

type logDispatcher struct {
	inputCh      chan []byte
	logRequestCh chan logRequest
	fullLog      []byte
	followers    []logResponseCh
}

func (l *logDispatcher) start() {
	for {
		select {
		case b := <-l.inputCh:
			l.fullLog = append(l.fullLog, b...)
			for _, follower := range l.followers {
				follower <- b
			}
			l.followers = nil
		case req := <-l.logRequestCh:
			if req.startIdx < len(l.fullLog) {
				req.ch <- l.fullLog[req.startIdx:]
			} else {
				l.followers = append(l.followers, req.ch)
			}
		}
	}
}

func readLogs(reqCh chan logRequest, w io.Writer) error {
	ch := make(logResponseCh)
	req := logRequest{startIdx: 0, ch: ch}
	reqCh <- req
	for b := range ch {
		req.startIdx += len(b)
		w.Write(b)
		reqCh <- req
	}
	return nil
}
```

nklaassen commented on proto/telejob.proto:
```diff
@@ -0,0 +1,84 @@
+  bool follow = 2;
+}
+
+// LogsResponse contains a chunk of logs.
+message LogsResponse {
+  bytes chunk = 1;
+}
```
are stdout and stderr combined into a single stream?

juliaogris replied:
Yes, they are combined, inspired by `docker logs`. 
I've added a comment in the proto definitions and the log streaming section in the design doc.

juliaogris commented:
@smallinsky , @nklaassen thank you very much for your detailed reviews.

I've addressed all your comments and pushed a single fixup commit which 
I will squash before merging.

nklaassen reviewed (approved):
Looks good!

r0mant commented on docs/design-doc.md:
```diff
@@ -0,0 +1,409 @@
+
+1. **Command Validation:** The server validates the provided command to ensure
+   it's not empty.
+2. **Job ID Generation:** A unique job ID is generated for the new job.
```
How are you planning to generate job ID?

juliaogris replied:
I was thinking of `1, 2, 3` generated by an `atomic.Uint64`, e.g.:
```go
id := strconv.FormatUint(maxID.Add(1), 10)
```
where `maxID atomic.Uint64` is a field of the type managing the jobs.
